### PR TITLE
feat(ui): Allow apps to register a custom seek bar UI implementation

### LINKED ISSUE #2924: Allow apps to register a custom seek bar UI element implementation
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes

**Is your feature request related to a problem? Please describe.**

Apps may want to create a custom SeekBar implementation, but this is currently unconfigurable.  See #2922 and specifically https://github.com/google/shaka-player/issues/2922#issuecomment-712430934

**Describe the solution you'd like**

1. Make an external interface to represent the public methods of `shaka.ui.SeekBar`, and make `SeekBar` an implementation of it
2. Offer a UI config with an element factory that apps can use to change the implementation of the seek bar (defaulting to `shaka.ui.SeekBar`).

**Describe alternatives you've considered**

There are workarounds, but they are not nice.  See https://github.com/google/shaka-player/issues/2922#issuecomment-712430934

### PR DESCRIPTION
**Pull request by hichem-code-it**
- Make an external interface shaka.extern.ISeekBar to represent the public methods of shaka.ui.SeekBar, and make SeekBar an implementation of it.
- Add a new function registerSeekBarFactory in shaka.ui.Controls  to register a seekBarFactory that will be used when present to create the SeekBar in the control elements.

Closes #2924

### CODE REVIEW

**Comment by hichem-code-it** [comment]
@googlebot I signed it!

File: ui/controls.js [hunk 1]
```diff
@@ -892,6 +892,36 @@ shaka.ui.Controls = class extends shaka.util.FakeEventTarget {
...
+   * This behaviour can be overriden by providing a SeekBar factory using the 
+   * registerSeekBarFactory function.
+   * 
+   * @private
+   */
+  addSeekBar_(){
+
+    if (this.config_.addSeekBar) {
+      
+      if( shaka.ui.ControlsPanel.seekBarFactory_ == null )
```
Comment (joeyparrish) [inline ui/controls.js:908]: Instead, please default the value of seekBarFactory_ to a factory associated with the default seek bar implementation.  That factory should be exported to allow apps to access it and re-register the original at any time.

File: ui/controls.js [hunk 2]
```diff
@@ -892,6 +892,36 @@ shaka.ui.Controls = class extends shaka.util.FakeEventTarget {
...
+      {
+        this.seekBar_ = new shaka.ui.SeekBar(this.bottomControls_, this);
+      }
+      else{
```
Comment (joeyparrish) [inline ui/controls.js:912]: There are many style violations the linter will complain about in this PR.  Please run `python build/check.py` to run the smallest set of linter and compiler checks, or `python build/all.py` to run all lint & build steps.

File: ui/externs/ui.js [hunk 3]
```diff
@@ -357,3 +357,49 @@ shaka.extern.IUISettingsMenu = class {
     this.backSpan;
   }
 };
+
+/**
+ * Interface for SeekBars. SeekBars should inherit from the concrete base
+ * class shaka.ui.Element.  
+ *
+ * @extends {shaka.extern.IUIElement }
```
Comment (joeyparrish) [inline ui/externs/ui.js:365]: style: remove the space inside the curly braces

File: ui/externs/ui.js [hunk 4]
```diff
@@ -357,3 +357,49 @@ shaka.extern.IUISettingsMenu = class {
...
+ * @interface
+ * @exportDoc
+ */
+shaka.extern.ISeekBar = class {
+  /**
+   * @param {!HTMLElement} parent
+   * @param {!shaka.ui.Controls} controls
+   */
+  constructor(parent, controls) {
+  }
```
Comment (joeyparrish) [inline ui/externs/ui.js:375]: style nit: move the closing curly brace to the previous line

File: ui/seek_bar.js [hunk 5]
```diff
@@ -191,7 +192,11 @@ shaka.ui.SeekBar = class extends shaka.ui.RangeElement {
     }
   }
 
-  /** @return {boolean} */
+  /**
+   *  @return {boolean}
```
Comment (joeyparrish) [inline ui/seek_bar.js:196]: When using `@override`, you don't need to specify the other jsdocs unless they differ from the base/interface.

File: ui/seek_bar.js [hunk 6]
```diff
@@ -200,6 +205,8 @@ shaka.ui.SeekBar = class extends shaka.ui.RangeElement {
   /**
    * Called by Controls on a timer to update the state of the seek bar.
```
Comment (joeyparrish) [inline ui/seek_bar.js:206]: These docs should go to the interface, so someone can understand the interface without referring to this implementation.

**Review by joeyparrish — CHANGES_REQUESTED** [review]
Thanks for your contributions!

File: ui/controls.js (hunk 1 above)
Comment (hichem-code-it) [reply ui/controls.js]: I had, that in mind, but somehow I was too lazy to create a new class : )

Now you gave me a good reason to do it

File: ui/controls.js (hunk 2 above)
Comment (hichem-code-it) [reply ui/controls.js]: My bad, it should be easy to fix !

File: ui/externs/ui.js (hunk 3 above)
Comment (hichem-code-it) [reply ui/externs/ui.js]: This one was not reported by the linter, but sure it's ugly now that you pointed it out : )

File: ui/externs/ui.js (hunk 4 above)
Comment (hichem-code-it) [reply ui/externs/ui.js]: At the end, this constructor was useless, I removed it.

File: ui/seek_bar.js (hunk 5 above)
Comment (hichem-code-it) [reply ui/seek_bar.js]: Done

File: ui/seek_bar.js (hunk 6 above)
Comment (hichem-code-it) [reply ui/seek_bar.js]: Done

File: ui/externs/ui.js [hunk 7]
```diff
@@ -395,7 +388,7 @@ shaka.extern.ISeekBar = class {
  * @interface
  * @exportDoc
  */
-shaka.extern.ISeekBar.SeekBarFactory = class {
+shaka.extern.ISeekBarFactory = class {
```
Comment (joeyparrish) [inline ui/externs/ui.js:391]: Please use the same convention as we did for IUIElement, IUIRangeElement, IUISettingsMenu.  That would be `shaka.extern.IUISeekBar` and `shaka.extern.IUISeekBar.Factory`.  Thanks!

File: ui/seek_bar_factory.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+goog.provide('shaka.ui.SeekBarFactory');
+
+goog.require('shaka.ui.SeekBar');
+
+/**
+ * @implements {shaka.extern.ISeekBarFactory}
+ * @export
+ */
+
+shaka.ui.SeekBarFactory = class {
```
Comment (joeyparrish) [inline ui/seek_bar_factory.js:16]: In keeping with convention, please call this `shaka.ui.SeekBar.Factory` and move it to the same file as `shaka.ui.SeekBar`.  Thanks!

**Review by joeyparrish — CHANGES_REQUESTED** [review]

File: ui/controls.js [hunk 9]
```diff
@@ -18,14 +18,14 @@ goog.require('shaka.ui.AdPosition');
...
 goog.require('shaka.ui.Localization');
-goog.require('shaka.ui.SeekBar');
 goog.require('shaka.ui.Utils');
 goog.require('shaka.util.Dom');
 goog.require('shaka.util.EventManager');
 goog.require('shaka.util.FakeEvent');
 goog.require('shaka.util.FakeEventTarget');
 goog.require('shaka.util.IDestroyable');
 goog.require('shaka.util.Timer');
+goog.require('shaka.ui.SeekBarFactory');
```
Comment (joeyparrish) [inline ui/controls.js:28]: When you rename this to `SeekBar.Factory`, you should put back `goog.require('shaka.ui.SeekBar')` and put it back in alphabetical order.

File: ui/controls.js [hunk 10]
```diff
@@ -353,6 +353,14 @@ shaka.ui.Controls = class extends shaka.util.FakeEventTarget {
     shaka.ui.ControlsPanel.elementNamesToFactories_.set(name, factory);
   }
 
+  /**
+   * @param {!shaka.extern.ISeekBarFactory} factory
+   * @export
+   */
+  static registerSeekBarFactory(factory) {
```
Comment (joeyparrish) [inline ui/controls.js:360]: Let's call this `registerSeekBar`, since the method next to it is `registerElement` and not `registerElementFactory`.

**Review by joeyparrish — CHANGES_REQUESTED** [review]

File: ui/externs/ui.js (hunk 7 above)
Comment (hichem-code-it) [reply ui/externs/ui.js]: Done

File: ui/seek_bar_factory.js (hunk 8 above)
Comment (hichem-code-it) [reply ui/seek_bar_factory.js]: Done

File: ui/controls.js (hunk 9 above)
Comment (hichem-code-it) [reply ui/controls.js]: Done

File: ui/controls.js (hunk 10 above)
Comment (hichem-code-it) [reply ui/controls.js]: Done ( even if registerFactory, makes more sens to me ). consistency, consistency : )

**Comment by DDeis** [comment]
Following the email I received from @hichem-code-it:

@googlebot I consent.

**Comment by DDeis** [comment]
@hichem-code-it, I did what you asked from your email, but I think this warning appears because for some commits you committed as `hichem` and not `hichem-code-it`. The email associated with those commit may not be the same as the one you added in `CONTRIBUTOR` and which you signed the CLA with ?

**Comment by hichem-code-it** [comment]
@DDeis  Thanks  ! I also corrected my wrong commit name ( for the merge I used the github interface directly, it seems like it is not configured correctly).

But the google bot still complains ! 

Maybe I also need the consent of @avelad who's commit was also included in my merge. @avelad can you please help me with that ?

**Comment by avelad** [comment]
@googlebot I consent.

**Pull request closed without merging**